From 91db3739e87280f35c92553725da5ff99a02f807 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:33 -0400
Subject: [infiniband] mthca: Return an error on ib_copy_to_udata() failure

Message-id: <b4a1d345b3e0ec92aec7784b64a8f962aa143012.1410425016.git.avadai@redhat.com>
Patchwork-id: 93767
O-Subject: [Patch RHEL7.1 V1 097/197] IB/mthca: Return an error on ib_copy_to_udata() failure
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Yann Droneaud <ydroneaud@opteya.com>

In case of error when writing to userspace, the function mthca_create_cq()
does not set an error code before following its error path.

This patch sets the error code to -EFAULT when ib_copy_to_udata() fails.

This was caught when using spatch (aka. coccinelle)
to rewrite call to ib_copy_{from,to}_udata().

Link: https://www.gitorious.org/opteya/coccib/source/75ebf2c1033c64c1d81df13e4ae44ee99c989eba:ib_copy_udata.cocci
Link: http://marc.info/?i=cover.1394485254.git.ydroneaud@opteya.com
Cc: <stable@vger.kernel.org>
Signed-off-by: Yann Droneaud <ydroneaud@opteya.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 08e74c4b00c30c232d535ff368554959403d0432)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mthca/mthca_provider.c b/drivers/infiniband/hw/mthca/mthca_provider.c
index 6440800..415f8e1 100644
--- a/drivers/infiniband/hw/mthca/mthca_provider.c
+++ b/drivers/infiniband/hw/mthca/mthca_provider.c
@@ -695,6 +695,7 @@ static struct ib_cq *mthca_create_cq(struct ib_device *ibdev, int entries,
 
  if (context && ib_copy_to_udata(udata, &cq->cqn, sizeof (__u32))) {
   mthca_free_cq(to_mdev(ibdev), cq);
+  err = -EFAULT;
   goto err_free;
  }
 
-- 
1.7.1